agreement/persistence.go

```diff
@@ -177,37 +193,72 @@ func restore(log logging.Logger, crash db.Accessor) (raw []byte, err error) {
 // decode process the incoming raw bytes array and attempt to reconstruct the agreement state objects.
 //
 // In all decoding errors, it returns the error code in err
-func decode(raw []byte, t0 timers.Clock, log serviceLogger) (t timers.Clock, rr rootRouter, p player, a []action, err error) {
+func decode(raw []byte, t0 timers.Clock, log serviceLogger, reflect bool) (t timers.Clock, rr rootRouter, p player, a []action, err error) {
 	var t2 timers.Clock
 	var rr2 rootRouter
 	var p2 player
 	a2 := []action{}
 	var s diskState
-
-	err = protocol.DecodeReflect(raw, &s)
-	if err != nil {
-		log.Errorf("decode (agreement): error decoding retrieved state (len = %v): %v", len(raw), err)
-		return
+	if reflect {
+		err = protocol.DecodeReflect(raw, &s)
+		if err != nil {
+			log.Errorf("decode (agreement): error decoding retrieved state (len = %v): %v", len(raw), err)
+			return
+		}
+	} else {
+		err = protocol.Decode(raw, &s)
+		if err != nil {
+			log.Warnf("decode (agreement): error decoding retrieved state using msgp (len = %v): %v. Trying reflection", len(raw), err)
+			err = protocol.DecodeReflect(raw, &s)
```

Comment:
What are the thoughts on either testing this or removing this? And if testing any suggestions for the approach?

The existing tests already guarantee that all structs that implement marshaller interfaces are testing using randomized objects in codec_tester.go so I can't think of a situation where Decode would fail but DecodeReflect would succeed without somebody bypassing codegen CI tests. Also any failures in automated tests were on things like nil pointer exceptions at test runtime due to optimism of the msgp library about objects that are outside of current package or mishandling things like *serializableError but all of those have been fixed.

Reply:
> And if testing any suggestions for the approach?

Yes, have a struct definition copied, add a new field, serialize, and deserialize into the original without this field. msgp will fail. Reflect would also fail tho but with an error and at least we can check log output.